lib/galaxy/webapps/galaxy/api/tools.py

```diff
@@ -426,6 +428,51 @@ def citations(self, trans: GalaxyWebTransaction, id, **kwds):
             rval.append(citation.to_dict('bibtex'))
         return rval
 
+    @expose_api
+    def conversion(self, trans: GalaxyWebTransaction, tool_id, payload, **kwd):
+        converter = self._get_tool(tool_id, user=trans.user)
+        target_type = payload.get("target_type")
+        source_type = payload.get("source_type")
+        input_src = payload.get("src")
+        input_id = payload.get("id")
+        # List of string of dependencies
+        try:
+            deps = trans.app.datatypes_registry.converter_deps[source_type][target_type]
+        except KeyError:
+            deps = {}
+        # Generate parameter dictionary
+        params = {}
+        # determine input parameter name and add to params
+        input_name = 'input1'
+        for key, value in converter.inputs.items():
+            if deps and value.name in deps:
+                params[value.name] = deps[value.name]
```

Comment:
It is getting the parameters for the converter; borrowed code from https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/datatypes/data.py#:~:text=def%20convert_dataset(self,value.type%20%3D%3D%20%27data%27%3A
which gets dependencies and parameters for the converter and then attempts to convert a dataset.

Reply:
I agree this piece of logic is sort of inscrutable - I've touched it many times but I still don't understand it - but I don't think it is the fault of this PR. I think what would make the review easier is if you de-duplicated that piece of code by adding a method to the tool class that built up that ``params`` dict. This would make reviewing the PR easier I think and prevent future divergence.

While you're at it - I suspect the following piece of code from the target code:

```
        # add potentially required/common internal tool parameters e.g. '__job_resource'
        if target_context:
            for key, value in target_context.items():
                if key.startswith('__'):
                    params[key] = value
```

is actually needed? I'm guessing this converter API won't work on instances with default job resources enabled. I'm unsure how to initialize target_content though. Having this piece of code shared between the two places though would make that easier to fix once we realize it is broken though.